apps/rush-lib/src/logic/operations/OperationExecutionManager.ts

```diff
@@ -208,182 +222,141 @@ export class OperationExecutionManager {
+       */
+      case OperationStatus.Failure:
+        // Failed operations get reported, even if silent.
+        // Generally speaking, silent operations shouldn't be able to fail, so this is a safety measure.
+        const message: string | undefined = record.error?.message;
+        if (message) {
+          terminal.writeStderrLine(message);
+        }
+        terminal.writeStderrLine(colors.red(`"${name}" failed to build.`));
+        const blockedQueue: Set<OperationExecutionRecord> = new Set(record.consumers);
+        for (const blockedRecord of blockedQueue) {
+          if (blockedRecord.status === OperationStatus.Ready) {
+            this._completedOperations++;
+
+            //
+            terminal.writeStdoutLine(`"${blockedRecord.name}" is blocked by "${name}".`);
+            blockedRecord.status = OperationStatus.Blocked;
+
+            for (const dependent of blockedRecord.consumers) {
+              blockedQueue.add(dependent);
+            }
+          }
+        }
+        this._hasAnyFailures = true;
+        break;
+      /**
+       * This operation was restored from the build cache.
+       */
+      case OperationStatus.FromCache:
+        if (!silent) {
+          terminal.writeStdoutLine(colors.green(`"${name}" was restored from the build cache.`));
+        }
+        break;
+      /**
+       * This operation was skipped via legacy change detection.
+       */
+      case OperationStatus.Skipped:
+        if (!silent) {
+          terminal.writeStdoutLine(colors.green(`"${name}" was skipped.`));
+        }
+        // Skipping means cannot guarantee integrity, so prevent cache writes in dependents.
+        blockCacheWrite = true;
+        break;
+      case OperationStatus.Success:
+        if (!silent) {
+          terminal.writeStdoutLine(
+            colors.green(`"${name}" completed successfully in ${record.stopwatch.toString()}.`)
+          );
+        }
+        // Legacy incremental build, if asked, prevent skip in dependents if the operation executed.
+        blockSkip ||= !this._changedProjectsOnly;
+        break;
+      case OperationStatus.SuccessWithWarning:
+        if (!silent) {
+          terminal.writeStderrLine(
+            colors.yellow(`"${name}" completed with warnings in ${record.stopwatch.toString()}.`)
+          );
+        }
+        // Legacy incremental build, if asked, prevent skip in dependents if the operation executed.
+        blockSkip ||= !this._changedProjectsOnly;
+        this._hasAnyNonAllowedWarnings = this._hasAnyNonAllowedWarnings || !runner.warningsAreAllowed;
+        break;
     }
-    operation.status = OperationStatus.Success;
 
-    operation.dependents.forEach((dependent: Operation) => {
-      if (!this._changedProjectsOnly) {
-        dependent.runner.isSkipAllowed = false;
+    // Apply status changes to direct dependents
+    for (const item of record.consumers) {
+      if (blockCacheWrite) {
```

Comment:
Would be good if we could do something like pass in an OperationCacheManager that could separately control cache management. Not a blocker since this is still in keeping with existing functionality, but a thought for the future.

Reply:
Will consider in the future as part of trying to separate out the cache into a plugin.